sql, opt: implements index skip scan

ridwanmsharif opened the pull request:
This change plans index-skip-scan in the optimizer.
It adds exploration rules to a scan but having them use the
loose index table reader instead of the regular table reader. 
This will allow the optimizer to create a far better plan (an 
order of magnitude better) for some `Distinct` queries by 
leveraging the index the scan is on. This will also make it 
possible to plan index-skip-scans more often when the
prefix of an index is known to have a small number of distinct 
values.

Commit:
sql: allow index-skip-scans to specify the prefix length

This change adds functionality to support index-skip-scans
when only a proper subset of the output columns are part
of the columns we do an index-skip scan over.

Release note: None

Commit:
opt: plan index skip scans

This change introduces index skip scans. To start with,
we implement index-skip scans when we have a DistinctOn
on a prefix of a scan.
Following this change, we can add more uses of this
when the DistinctOn has an index-join as its input,
when there is no DistinctOn but the prefix of the index
has very few distinct values and there is a constraint
on the columns after those columns (This would need
some more thought but I think it's possible if we
do an index skip scan and a cross product with the other
bounded column values and use this cross product to do a
lookup join with the original table).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39668)
<!-- Reviewable:end -->

ridwanmsharif commented:
The following commits add a few things that are missing now. For example, performing index-skip-scan using a secondary index that needs an index join.

Also, I'm unfamiliar with the DistSQL planning, and have doubts about whether we can remove the `DistintOn` and `AggDistint` in these transformations. I'm imagining a case where we have separate nodes that have rows with the same prefix. I think each node would then read those rows and then output them. If we remove the `DistinctOn`/`AggDistinct` (as I'm doing in the PR now), we may have an incorrect result with multiple rows with the same `Distinct` cols

Commit:
opt: plan index skip scans on secondary indexes through an index join

This change adds an exploration rule that allows
index skip scans to be used when we have an index
join underneath a `DistinctOn`.

Release note: None

Commit:
opt: add operator for index skip scans

Previously we were planning an index skip scan as a regular scan
with some other properties. This change separates it out into its own
operator. This is good because even though a lot of rules that apply to
scans are applicable here too, some of them are difficult to reason
about and separating it into a separate operator allows us to audit
those rules in isolation before applying them instead of just doing it
now. This also helps when creating new rules in the future.

Another important note is, we need to do some careful work with
IndexSkipScan so the exploration rules works sanely when the data is
distributed over multiple nodes. This is important because when each
node does an `IndexSkipScan` over its own data, the rows returned by
different nodes might conflict and so we need a `DistinctOn` to resolve
them. and We add this DistinctOn node in the exec builder phase so the
guarantees made by the operator always hold.

Finally, this change tries to IndexSkipScan behaves sanely
when used with various different orderings.

Release note: None

ridwanmsharif commented:
FYI if you're reviewing already, I'm adding a change that introduces a new `IndexSkipScan` operator so we can make the guarantees about the output more explicit. This way all the rules that apply to `Scan`s aren't directly applied to `IndexSkipScan`s too before we audit if its appropriate. This way we can remove the logic of adding the `DistinctOn` from the rule itself and instead add it in the exec factory.

ridwanmsharif commented:
I thought this was going to be a nice quick improvement but this got out of hand a bit too quickly, sorry for that. There were a few complications with this but they seem to have all been addressed now. There are a few high-level things to note about this PR that need some careful thought:

- An index skip scan operator guarantees for each unique index prefix, we will return 1 value. This guarantee is tricky because if you have data on multiple nodes, each node will use the index skip scan table reader on their local spans but the results might not be unique globally. As a result, to provide this guarantee we either have to add a distinct on in the xform rules or add a distinct on during the exec building phase (I've implemented the latter so the rules don't have to worry about data being distributed across nodes)
- An index join can't have an index skip scan as input. Its just not something we can support yet. The reason for this is because we implement an index skip scan as distinct on and a scan that skips and as a result (because the input to an index join can't take a distinct node) this PR introduces a workaround that still tries to use an index skip scan in that case. It does so by converting the index join into a lookup join.
- When the `DIstinctOn` has an input ordering, this ordering can affect what rows we see in the index skip scan (because based on the ordering we maybe skipping some rows that we shouldn't).
- I found working with the ordering is a little tricky and making it play well with enforcers were a little challenging. But the current solution does seem to work (see the logictests in the PR). What we do is, we specify the ordering of the index skip scan when we create it. We also specify whether the index skip scan scans in reverse.
- The `DistinctOn` returns a subset of columns of its input, so when we replace the `DistinctOn`, there is some pruning logic there that may need some love.

I will try and respond to questions and might also try and fixup this PR based on the reviews. For the time being, I'll create issues tracking some other places we can use this PR so to motivate getting this in soon.

RaduBerinde reviewed (commented):
I think that from the optimizer side, the operator should be semantically equivalent to a DistinctOn -> Scan complex, where:
 - the distinct columns are a prefix of the index
 - the ordering inside Distinct refers only to distinct columns
 - the constraints in the scan only cover the distinct columns

Some of these could be relaxed a bit in the future if necessary, but with non-trivial work.

The distsql physical planner should be in charge of adding distinct processors as needed to guarantee these semantics in distributed cases.

Separately, we should fix the restriction of `indexJoin` input nodes, now that the local execution code is gone it should be doable. I'm surprised we got this far without fixing that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39668#-:-LmyzU7DB0zmdp2_n6Nm:bagcrct)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj)



<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
Hmm those are good points. Right now the `IndexSkipScan` is equivalent to a DistinctOn - > Scan (we add the DistinctOn when building the exec plan). Also yes, the distinct on has to be on the prefix of the index currently.

> - the ordering inside Distinct refers only to distinct columns
> - the constraints in the scan only cover the distinct columns

The PR doesn't restrict either of these to just the distinct columns actually. Why do we need these restrictions exactly?
There are tests I added that tests `IndexSkipScan`s work sanely even when the ordering is on columns that are part of the index but not part of the distinct on. Same thing with the constraints, why do you think the restriction on the distinct cols is necessary?

> The distsql physical planner should be in charge of adding distinct processors as needed to guarantee these semantics in distributed cases.

I agree with this. Until this is true though, the `DistinctOn` added in the exec build phase is a workaround I'm okay with.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39668#-:-LmzGND1AesT90brWD5M:byv8wid)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> The PR doesn't restrict either of these to just the distinct columns actually. Why do we need these restrictions exactly?
There are tests I added that tests IndexSkipScans work sanely even when the ordering is on columns that are part of the index but not part of the distinct on. Same thing with the constraints, why do you think the restriction on the distinct cols is necessary?

They're not strictly speaking necessary (I did mention that we could relax them), they're just hard to think about and get right IMO. If we have an ordering on other columns, we have to set up the distinct processors with ordered input synchronizers to select the correct row. There may be cases where we're not even selecting those other columns, but we would need to get the values over to the distinct processor.

Constraints may not be a problem, it's more that they normally come from filters which in general can't support on those other columns anyway.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39668#-:-LmzIffr8DcOm6rzDRZ-:b-bj6dyh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj)



<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
Yeah that's fair. We could tighten the requirements for this rule a little more than we do in this PR and then loosen it later. For now, I've added support for orderings that aren't part of the `DistinctOn` columns as well and have tests exercising behavior in those situations.

I'll defer to the team from here! Thanks for letting me work on such cool projects!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39668#-:-Ln-5DMY0aeBoYl_7xKe:bp33su9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj)



<!-- Sent from Reviewable.io -->

andreimatei commented:
https://github.com/cockroachdb/cockroach/pull/51178 deleted the `indexSkipTableReader` processor cause it was laying around unused, so this PR needs to revert the deletion when the time comes.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=39668) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=39668) before we can accept your contribution.<br/><hr/>**Ridwan Sharif** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=39668) it.</sub>
